opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -1538,88 +1534,26 @@ namespace Opm {
                         }
 
                         well_controls_iset_distr(ctrl, rctrl, & distr[0]);
-                    }
-                }
-
-                // RESV control, WCONHIST wells.  A bit of duplicate
-                // work, regrettably.
-                if (is_producer && wells()->name[*rp] != 0) {
-                    WellMap::const_iterator i = wmap.find(wells()->name[*rp]);
-
-                    if (i != wmap.end()) {
-                        const auto* wp = i->second;
 
-                        const WellProductionProperties& p =
-                            wp->getProductionProperties(step);
+                        // for the WCONHIST wells, we need to calculate the RESV rates since it can not be specified directly
+                        if (is_producer) {
+                            const WellMap::const_iterator i = wmap.find(wells()->name[*rp]);
 
-                        if (! p.predictionMode) {
-                            // History matching (WCONHIST/RESV)
-                            SimFIBODetails::historyRates(pu, p, hrates);
-
-                            const int fipreg = 0; // Hack.  Ignore FIP regions.
-                            rateConverter_->calcCoeff(fipreg, pvtreg, distr);
+                            if (i == wmap.end()) {
+                                OPM_THROW(std::runtime_error, "Failed to find the well " << wells()->name[*rp] << " in wmap.");
+                            }
+                            const auto* wp = i->second;
+                            const WellProductionProperties& production_properties = wp->getProductionProperties(step);
+                            // historical phase rates
+                            std::vector<double> hrates(np);
+                            SimFIBODetails::historyRates(phase_usage_, production_properties, hrates);
 
-                            // WCONHIST/RESV target is sum of all
-                            // observed phase rates translated to
-                            // reservoir conditions.  Recall sign
-                            // convention: Negative for producers.
                             std::vector<double> hrates_resv(np);
                             rateConverter_->calcReservoirVoidageRates(fipreg, pvtreg, hrates, hrates_resv);
```

Comment:
I do not re-call why we use different ways in obtaining the RESV voidage rate when calculating the RESV target and when calculating RESV rate when checking `constraintBroken()` with `rateToCompare()` . 

We need slightly more sophisticated techniques to calculate the RESV target with  calcReservoirVoidageRates(), while we use \Sum_p distr[p] * rates[p] when we need to compare `rates` and the target.

Reply:
In any way it is an inconsistency, which might hide some potential problems and make the debugging tricky.